pysmt/simplifier.py

```diff
@@ -288,44 +288,37 @@ def walk_plus(self, formula, args, **kwargs):
         return self.manager.Plus(ns)
 
     def walk_times(self, formula, args, **kwargs):
-        assert len(args) == 2
-
-        sl = args[0]
-        sr = args[1]
-
-        if sl.is_constant() and sr.is_constant():
-            l = sl.constant_value()
-            r = sr.constant_value()
-            if sl.is_real_constant():
-                return self.manager.Real(l * r)
-            elif sl.is_int_constant():
-                return self.manager.Int(l * r)
+        if all(x.is_constant() for x in args):
+            mul = args[0].constant_value()
+            for x in args[1:]:
+                mul *= x.constant_value()
+            if args[0].is_real_constant():
+                return self.manager.Real(mul)
+            elif args[0].is_int_constant():
+                return self.manager.Int(mul)
             else:
-                assert sl.is_algebraic_constant()
+                assert args[0].is_algebraic_constant()
                 from pysmt.numeral import Numeral
-                return self.manager._Algebraic(Numeral(l * r))
-
-        if sl.is_constant():
-            if sl.is_one():
-                return sr
-            elif sl.is_zero():
-                if sl.is_real_constant():
-                    return self.manager.Real(0)
-                else:
-                    assert sl.is_int_constant()
-                    return self.manager.Int(0)
-
-        if sr.is_constant():
-            if sr.is_one():
-                return sl
-            elif sr.is_zero():
-                if sr.is_real_constant():
-                    return self.manager.Real(0)
-                else:
-                    assert sr.is_int_constant()
-                    return self.manager.Int(0)
+                return self.manager._Algebraic(Numeral(mul))
+
+        new_args = []
```

Comment:
I think we should unify both cases. In particular, we iterate on the arguments and use an accumulator for the constants, while maintaining the non-constant objects. Currently you are doing 2 passes of the arguments. I think you can do it in one.

* If it is zero: return 0
* If it is one: continue
* If is is a constant: add to accumulator
* If is is not a constant, add to new_args

Do you have anther reason for this?